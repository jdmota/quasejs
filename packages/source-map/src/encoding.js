const win = typeof window !== "undefined" && window; // eslint-disable-line
const encoding = {};

if ( win && typeof win.btoa === "function" ) {
  encoding.encode = win.btoa;
  encoding.decode = win.atob;
} else if ( typeof Buffer === "function" ) {
  encoding.encode = str => Buffer.from( str ).toString( "base64" );
  encoding.decode = str => Buffer.from( str, "base64" ).toString();
}

export default encoding;
