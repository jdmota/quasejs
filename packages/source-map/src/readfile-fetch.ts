export class GetFile {

  private fileCache: Map<string, Promise<string>>;
  private fetchCache: Map<string, Promise<Response>>;
  private window: Window | false;
  private fs: {
    readFile( path: string, encoding: string, callback: ( err: NodeJS.ErrnoException | null, data: string ) => void ): void;
  } | null;

  constructor() {
    this.fileCache = new Map();
    this.fetchCache = new Map();
    this.window = typeof window !== "undefined" && window; // eslint-disable-line no-undef
    this.fs = require( "fs" );
  }

  private _fetchResponse( file: string ) {
    const { window } = this;
    if ( !window ) {
      throw new Error( `No window` );
    }
    return window.fetch( file );
  }

  private _readFile( file: string ) {
    const { fs } = this;
    if ( !fs ) {
      throw new Error( `No fs` );
    }
    return new Promise<string>( ( resolve, reject ) => {
      fs.readFile( file, "utf8", ( err, result ) => {
        if ( err ) {
          reject( err );
        } else {
          resolve( result );
        }
      } );
    } );
  }

  fetchResponse( file: string ) {
    let curr = this.fetchCache.get( file );
    if ( curr ) {
      return curr;
    }
    curr = this._fetchResponse( file );
    this.fetchCache.set( file, curr );
    return curr;
  }

  async fetch( file: string ) {
    const r = await this.fetchResponse( file );
    return r.text();
  }

  readFile( file: string ) {
    let curr = this.fileCache.get( file );
    if ( curr ) {
      return curr;
    }
    curr = this._readFile( file );
    this.fileCache.set( file, curr );
    return curr;
  }

  delete( file: string ) {
    this.fileCache.delete( file );
    this.fetchCache.delete( file );
  }

}
