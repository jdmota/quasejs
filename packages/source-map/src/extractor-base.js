// @flow
import encoding from "./encoding";

const { SourceMapConsumer } = require( "source-map" );
const { resolveAsUrl } = require( "@quase/path-url" );

const baseRegex = "\\s*[@#]\\s*sourceMappingURL\\s*=\\s*([^\\s]*)",
    // Matches /* ... */ comments
    regex1 = new RegExp( "/\\*" + baseRegex + "\\s*\\*/" ),
    // Matches // .... comments
    regex2 = new RegExp( "//" + baseRegex + "($|\n|\r\n?)" ),
    // Matches DataUrls
    regexDataUrl = /data:[^;\n]+;base64,(.*)/;

export type SourceMapInfo = {
  map: ?Object,
  mapLocation: string
};

export type SourceMapInfoWithMap = {
  map: Object,
  mapLocation: string
};

export type Original = {
  originalFile: string,
  originalCode: ?string
};

export default class SourceMapExtractorBase {

  static consumeSourceMap( data: string ): Object {
    return new SourceMapConsumer( JSON.parse( data ) );
  }

  getMapFromUrl( fileLocation: string, url: string ): SourceMapInfo {

    const dataUrlMatch = url.match( regexDataUrl );

    if ( dataUrlMatch ) {
      return {
        map: SourceMapExtractorBase.consumeSourceMap( encoding.decode( dataUrlMatch[ 1 ] ) ),
        mapLocation: fileLocation
      };
    }

    return {
      map: null,
      mapLocation: resolveAsUrl( fileLocation, url )
    };
  }

  getMapFromFile( fileLocation: string, fileContent: string ): ?SourceMapInfo {

    const match = fileContent.match( regex1 ) || fileContent.match( regex2 );

    if ( !match ) {
      return;
    }

    return this.getMapFromUrl( fileLocation, match[ 1 ] );
  }

  // generated: { line, column, bias? }
  getOriginalLocationFromMap( map: Object, mapLocation: string, generated: Object ): ?Original {

    const pos = map.originalPositionFor( generated ); // { source, line, column, name }

    if ( pos.line != null ) {
      const originalFile = resolveAsUrl( mapLocation, pos.source );
      const originalCode = map.sourceContentFor( pos.source, true );

      pos.originalFile = originalFile;
      pos.originalCode = originalCode;
      return pos;
    }

    return null;
  }

}
