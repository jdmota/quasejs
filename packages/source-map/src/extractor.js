import FileSystem from "../../fs/memory-fs/src";
import { makeAbsolute, resolveAsUrl } from "../../pathname/src/path-url";
import encoding from "./encoding";

const { SourceMapConsumer } = require( "source-map" );

const baseRegex = "\\s*[@#]\\s*sourceMappingURL\\s*=\\s*([^\\s]*)",
    // Matches /* ... */ comments
    regex1 = new RegExp( "/\\*" + baseRegex + "\\s*\\*/" ),
    // Matches // .... comments
    regex2 = new RegExp( "//" + baseRegex + "($|\n|\r\n?)" ),
    // Matches DataUrls
    regexDataUrl = /data:[^;\n]+;base64,(.*)/;

export default class SourceMapExtractor {

  constructor( fs = new FileSystem() ) {
    this.fs = fs;
    this.cacheMapLocation = Object.create( null );
    this.mapRequest = Object.create( null );
  }

  static consumeSourceMap( data ) {
    return data == null ? null : new SourceMapConsumer( JSON.parse( data ) );
  }

  async _getMap( file ) {

    const code = await this.fs.getFile( file );

    const match = code.match( regex1 ) || code.match( regex2 );

    if ( match ) {
      const url = match[ 1 ];
      const dataUrlMatch = url.match( regexDataUrl );

      if ( dataUrlMatch ) {
        return {
          map: SourceMapExtractor.consumeSourceMap( encoding.decode( dataUrlMatch[ 1 ] ) ),
          mapLocation: file
        };
      }

      const mapLocation = this.cacheMapLocation[ file ] = resolveAsUrl( file, url );

      let sourcemap;

      try {
        sourcemap = await this.fs.getFile( mapLocation );
      } catch ( e ) {
        // The sourcemap that was supposed to exist, was not found
      }

      return {
        map: SourceMapExtractor.consumeSourceMap( sourcemap ),
        mapLocation
      };
    }

  }

  getMap( file ) {
    file = makeAbsolute( file );
    return this.mapRequest[ file ] || ( this.mapRequest[ file ] = this._getMap( file ) );
  }

  purge( file ) {
    if ( file ) {
      file = makeAbsolute( file );
      this.fs.purge( file );
      this.mapRequest[ file ] = null;
      this.cache[ this.cacheMapLocation[ file ] ] = null;
      this.cacheMapLocation[ file ] = null;
    }
  }

  // file, generated: { line, column, bias? }
  async getOriginalLocation( file, generated ) {

    file = makeAbsolute( file );

    const { map, mapLocation } = await this.getMap( file ) || {};

    if ( map ) {
      const pos = map.originalPositionFor( generated ); // { source, line, column, name }

      if ( pos.line != null ) {
        const originalFile = resolveAsUrl( mapLocation, pos.source );
        const originalCode = map.sourceContentFor( pos.source, true ) || await this.fs.getFile( originalFile );

        pos.originalFile = originalFile;
        pos.originalCode = originalCode;

        return pos;
      }
    }

    return { code: await this.fs.getFile( file ) };
  }

}
