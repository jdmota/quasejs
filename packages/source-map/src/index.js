import encoding from "./encoding";
import SourceMapExtractor from "./extractor";
import SourceMapExtractorBase from "./extractor-base";

const { SourceMapConsumer, SourceMapGenerator } = require( "source-map" );

function getConsumer( map ) {
  if ( typeof map.generatedPositionFor === "function" ) {
    return map;
  }
  return new SourceMapConsumer( map );
}

export { SourceMapExtractor, SourceMapExtractorBase };
export { SourceMapConsumer, SourceMapGenerator };

export function joinSourceMaps( maps ) {

  maps = maps.filter( Boolean );

  if ( maps.length === 0 ) {
    return null;
  }

  if ( maps.length === 1 ) {
    return Object.assign( {}, maps[ 0 ] );
  }

  const inputMapConsumer = getConsumer( maps[ 0 ] );

  const mergedGenerator = new SourceMapGenerator( {
    file: inputMapConsumer.file,
    sources: inputMapConsumer.sources,
    sourceRoot: inputMapConsumer.sourceRoot,
    sourcesContent: inputMapConsumer.sourcesContent
  } );

  inputMapConsumer.eachMapping( m => {

    let pos = {
      line: m.generatedLine,
      column: m.generatedColumn
    };

    for ( let i = 1; i < maps.length; i++ ) {

      pos = getConsumer( maps[ i ] ).generatedPositionFor( {
        source: inputMapConsumer.file,
        line: pos.line,
        column: pos.column
      } );

      if ( pos.column == null ) {
        return;
      }

    }

    const newMapping = {
      source: null,
      original: null,
      name: m.name,
      generated: pos
    };

    if ( m.source && typeof m.originalLine === "number" && typeof m.originalColumn === "number" ) {
      newMapping.source = m.source;
      newMapping.original = { line: m.originalLine, column: m.originalColumn };
    }

    mergedGenerator.addMapping( newMapping );

  } );

  const map = Object.assign( {}, maps[ 0 ] );
  map.mappings = mergedGenerator.toJSON().mappings;
  return map;
}

export function sourceMapToString( map ) {
  return JSON.stringify( map );
}

export function sourceMapToUrl( map ) {
  if ( !encoding.encode ) {
    throw new Error( "Unsupported environment: `window.btoa` or `Buffer` should be supported." );
  }
  return "data:application/json;charset=utf-8;base64," + encoding.encode( sourceMapToString( map ) );
}

export const SOURCE_MAP_URL = "source" + "MappingURL"; // eslint-disable-line

export function sourceMapComment( url ) {
  return `\n//# ${SOURCE_MAP_URL}=${url}`;
}

export function getOriginalLocation( map, generated ) { // map, generated: { line, column, bias? }
  return getConsumer( map ).originalPositionFor( generated ); // { source, line, column, name }
}

