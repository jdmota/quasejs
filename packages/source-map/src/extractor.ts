import SourceMapExtractorBase, { Original, SourceMapInfoWithMap } from "./extractor-base";
import { GetFile } from "./readfile-fetch";
import { makeAbsolute, resolveAsUrl, isUrl } from "@quase/path-url";

export default class SourceMapExtractor extends SourceMapExtractorBase {

  private fileGetter: GetFile;
  private mapRequest: Map<string, Promise<SourceMapInfoWithMap | null>>;

  constructor() {
    super();
    this.fileGetter = new GetFile();
    this.mapRequest = new Map();
  }

  private read( file: string ): Promise<string> {
    return isUrl( file ) ? this.fileGetter.fetch( file ) : this.fileGetter.readFile( file );
  }

  private async _getMap( file: string ): Promise<SourceMapInfoWithMap | null> {

    const headers = isUrl( file ) ? ( await this.fileGetter.fetchResponse( file ) ).headers : null;
    const url = headers ? headers.get( "SourceMap" ) || headers.get( "X-SourceMap" ) : null;
    let mapData;

    if ( url ) {
      mapData = this.getMapFromUrl( file, url );
    } else {
      mapData = this.getMapFromFile( file, await this.read( file ) );
    }

    if ( !mapData ) {
      return null;
    }

    const mapLocation = mapData.mapLocation;
    const map = mapData.map == null ?
      JSON.parse( await this.read( mapData.mapLocation ) ) :
      mapData.map;

    return {
      map,
      mapLocation
    };
  }

  getMap( file: string ): Promise<SourceMapInfoWithMap | null> {
    file = makeAbsolute( file );

    let job = this.mapRequest.get( file );
    if ( !job ) {
      job = this._getMap( file );
      this.mapRequest.set( file, job );
    }
    return job;
  }

  purge( file: string ) {
    file = makeAbsolute( file );
    this.fileGetter.delete( file );
    this.mapRequest.delete( file );
  }

  async getOriginalLocation( file: string, generated: { line: number; column: number; bias?: number } ): Promise<Original | { code: string }> {

    const info = await this.getMap( file );

    if ( info && info.map ) {
      const { map, mapLocation } = info;
      const original = await this.getOriginalLocationFromMap( map, mapLocation, generated );

      if ( original ) {
        if ( original.originalCode == null && original.originalFile ) {
          original.originalCode = await this.read( original.originalFile );
        }
        return original;
      }
    }

    return {
      code: await this.read( file )
    };
  }

  async getOriginalSources( file: string ): Promise<string[]> {

    const info = await this.getMap( file );

    if ( info && info.map ) {
      const { map, mapLocation } = info;
      return map.sources.map( ( s: string ) => resolveAsUrl( mapLocation, s ) );
    }

    return [];
  }

}
