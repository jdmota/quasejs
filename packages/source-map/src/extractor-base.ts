import encoding from "./encoding";
import { RawSourceMap, RawIndexMap, SourceMapConsumer } from "source-map";

const { resolveAsUrl } = require( "@quase/path-url" );

const baseRegex = "\\s*[@#]\\s*sourceMappingURL\\s*=\\s*([^\\s]*)",
  // Matches /* ... */ comments
  regex1 = new RegExp( "/\\*" + baseRegex + "\\s*\\*/" ),
  // Matches // .... comments
  regex2 = new RegExp( "//" + baseRegex + "($|\n|\r\n?)" ),
  // Matches DataUrls
  regexDataUrl = /data:[^;\n]+;base64,(.*)/;

export type SourceMapType = string | RawSourceMap | RawIndexMap;

export type SourceMapInfo = {
  map: RawSourceMap | null;
  mapLocation: string;
};

export type SourceMapInfoWithMap = {
  map: RawSourceMap;
  mapLocation: string;
};

export type Original = {
  source: string | null;
  line: number | null;
  column: number | null;
  name: string | null;
  originalFile: string | null;
  originalCode: string | null;
};

export default class SourceMapExtractorBase {

  getMapFromUrl( fileLocation: string, url: string ): SourceMapInfo {

    const dataUrlMatch = url.match( regexDataUrl );

    if ( dataUrlMatch ) {
      return {
        map: JSON.parse( encoding.decode( dataUrlMatch[ 1 ] ) ),
        mapLocation: fileLocation
      };
    }

    return {
      map: null,
      mapLocation: resolveAsUrl( fileLocation, url )
    };
  }

  getMapFromFile( fileLocation: string, fileContent: string ): SourceMapInfo | null {

    const match = fileContent.match( regex1 ) || fileContent.match( regex2 );

    if ( !match ) {
      return null;
    }

    return this.getMapFromUrl( fileLocation, match[ 1 ] );
  }

  async getOriginalLocationFromMap( map: RawSourceMap | SourceMapConsumer, mapLocation: string, generated: { line: number; column: number; bias?: number } ): Promise<Original | null> {

    const consumer = map instanceof SourceMapConsumer ? map : await new SourceMapConsumer( map );

    const pos: {
      source: string | null;
      line: number | null;
      column: number | null;
      name: string | null;
    } = consumer.originalPositionFor( generated );

    let result;

    if ( pos.line != null ) {
      const originalFile = resolveAsUrl( mapLocation, pos.source );
      const originalCode = pos.source ? consumer.sourceContentFor( pos.source, true ) : null;

      result = {
        ...pos,
        originalFile,
        originalCode
      };
    } else {
      result = null;
    }

    if ( !( map instanceof SourceMapConsumer ) ) {
      consumer.destroy();
    }

    return result;
  }

}
