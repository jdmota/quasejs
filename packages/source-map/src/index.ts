import encoding from "./encoding";
import { SourceMapExtractor } from "./extractor";
import SourceMapExtractorBase, { SourceMapType } from "./extractor-base";
import { SourceMapConsumer, SourceMapGenerator, Position } from "source-map";

export { SourceMapExtractor, SourceMapExtractorBase };
export { SourceMapConsumer, SourceMapGenerator };

export async function joinSourceMaps(maps: SourceMapType[]) {
  if (maps.length === 0) {
    throw new Error(`No sourcemaps provided`);
  }

  if (maps.length === 1) {
    throw new Error(`Only 1 sourcemap provided`);
  }

  const consumer0 = await new SourceMapConsumer(maps[0]);
  const mergedGenerator = SourceMapGenerator.fromSourceMap(consumer0);
  consumer0.destroy();

  for (const map of maps.slice(1)) {
    const consumer = await new SourceMapConsumer(map);
    mergedGenerator.applySourceMap(consumer);
    consumer.destroy();
  }

  return mergedGenerator.toJSON();
}

export function sourceMapToString(map: SourceMapType) {
  return JSON.stringify(map);
}

export function sourceMapToUrl(map: SourceMapType) {
  if (!encoding.encode) {
    throw new Error(
      "Unsupported environment: `window.btoa` or `Buffer` should be supported."
    );
  }
  return (
    "data:application/json;charset=utf-8;base64," +
    encoding.encode(sourceMapToString(map))
  );
}

export const SOURCE_MAP_URL = "source" + "MappingURL"; // eslint-disable-line

export function sourceMapComment(url: string) {
  return `\n//# ${SOURCE_MAP_URL}=${url}`;
}

export function getOriginalLocation(
  map: SourceMapType,
  generated: Position & { bias?: number }
) {
  return SourceMapConsumer.with(map, null, consumer => {
    return consumer.originalPositionFor(generated);
  });
}
