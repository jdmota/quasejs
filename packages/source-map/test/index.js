import { SourceMapExtractor, SourceMapExtractorBase } from "../src";

const fs = require( "fs-extra" );
const path = require( "path" );

function relative( a, b ) {
  return path.relative( a, b ).replace( /\\/g, "/" );
}

test( "get map from file content", async() => {

  const extractor = new SourceMapExtractorBase();
  const fixturesFolder = path.join( __dirname, "fixtures" );
  const files = await fs.readdir( fixturesFolder );

  for ( const file of files ) {
    if ( /\.map$/.test( file ) ) {
      continue;
    }
    const fileLocation = path.join( __dirname, "fixtures", file );
    const info = extractor.getMapFromFile(
      fileLocation,
      await fs.readFile( fileLocation, "utf8" )
    );

    info.mapLocation = relative( fixturesFolder, info.mapLocation );
    expect( info ).toMatchSnapshot();
  }

} );

test( "get map from file", async() => {

  const extractor = new SourceMapExtractor( fs );
  const fixturesFolder = path.join( __dirname, "fixtures" );
  const files = await fs.readdir( fixturesFolder );

  for ( const file of files ) {
    if ( /\.map$/.test( file ) ) {
      continue;
    }
    const fileLocation = path.join( __dirname, "fixtures", file );
    const info = await extractor.getMap( fileLocation );

    info.mapLocation = relative( fixturesFolder, info.mapLocation );
    expect( info ).toMatchSnapshot();
  }

} );

test( "get original location from map", async() => {

  const extractor = new SourceMapExtractorBase();
  const fixturesFolder = path.join( __dirname, "fixtures" );
  const mapLocation = path.join( fixturesFolder, "map-file-comment.css.map" );
  const map = SourceMapExtractorBase.consumeSourceMap( await fs.readFile( mapLocation, "utf8" ) );

  const location = extractor.getOriginalLocationFromMap( map, mapLocation, {
    line: 8,
    column: 0
  } );

  location.originalFile = relative( fixturesFolder, location.originalFile );
  expect( location ).toMatchSnapshot();

} );
