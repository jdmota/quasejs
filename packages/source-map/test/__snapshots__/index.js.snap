// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get map from file 1`] = `
Object {
  "map": Object {
    "file": "map-file-comment.css",
    "mappings": "AAAA,wBAAyB;EACvB,UAAU,EAAE,IAAI;EAChB,MAAM,EAAE,KAAK;EACb,OAAO,EAAE,IAAI;EACb,aAAa,EAAE,iBAAiB;EAChC,KAAK,EAAE,OAAkB;;AAG3B,wBAAyB;EACvB,OAAO,EAAE,IAAI;;ACTf,gBAAiB;EACf,UAAU,EAAE,IAAI;EAChB,KAAK,EAAE,MAAM;;AAGf,kBAAmB;EACjB,MAAM,EAAE,IAAI;EACZ,OAAO,EAAE,IAAI;EACb,UAAU,EAAE,KAAK;EACjB,aAAa,EAAE,GAAG;EAClB,KAAK,EAAE,KAAK;;AAEd,kBAAmB;EACjB,KAAK,EAAE,KAAK;;AAGd,mBAAoB;EAClB,KAAK,EAAE,KAAK;EACZ,MAAM,EAAE,IAAI;EACZ,OAAO,EAAE,IAAI;EACb,SAAS,EAAE,IAAI",
    "sources": Array [
      "./client/sass/core.scss",
      "./client/sass/main.scss",
    ],
    "version": "3",
  },
  "mapLocation": "multi-line-inline.css",
}
`;

exports[`get map from file 2`] = `
Object {
  "map": Object {
    "file": "map-file-comment.css",
    "mappings": "AAAA,wBAAyB;EACvB,UAAU,EAAE,IAAI;EAChB,MAAM,EAAE,KAAK;EACb,OAAO,EAAE,IAAI;EACb,aAAa,EAAE,iBAAiB;EAChC,KAAK,EAAE,OAAkB;;AAG3B,wBAAyB;EACvB,OAAO,EAAE,IAAI;;ACTf,gBAAiB;EACf,UAAU,EAAE,IAAI;EAChB,KAAK,EAAE,MAAM;;AAGf,kBAAmB;EACjB,MAAM,EAAE,IAAI;EACZ,OAAO,EAAE,IAAI;EACb,UAAU,EAAE,KAAK;EACjB,aAAa,EAAE,GAAG;EAClB,KAAK,EAAE,KAAK;;AAEd,kBAAmB;EACjB,KAAK,EAAE,KAAK;;AAGd,mBAAoB;EAClB,KAAK,EAAE,KAAK;EACZ,MAAM,EAAE,IAAI;EACZ,OAAO,EAAE,IAAI;EACb,SAAS,EAAE,IAAI",
    "sources": Array [
      "./client/sass/core.scss",
      "./client/sass/main.scss",
    ],
    "version": "3",
  },
  "mapLocation": "map-file-comment.css.map",
}
`;

exports[`get map from file 3`] = `
Object {
  "map": Object {
    "file": "map-file-comment.css",
    "mappings": "AAAA,wBAAyB;EACvB,UAAU,EAAE,IAAI;EAChB,MAAM,EAAE,KAAK;EACb,OAAO,EAAE,IAAI;EACb,aAAa,EAAE,iBAAiB;EAChC,KAAK,EAAE,OAAkB;;AAG3B,wBAAyB;EACvB,OAAO,EAAE,IAAI;;ACTf,gBAAiB;EACf,UAAU,EAAE,IAAI;EAChB,KAAK,EAAE,MAAM;;AAGf,kBAAmB;EACjB,MAAM,EAAE,IAAI;EACZ,OAAO,EAAE,IAAI;EACb,UAAU,EAAE,KAAK;EACjB,aAAa,EAAE,GAAG;EAClB,KAAK,EAAE,KAAK;;AAEd,kBAAmB;EACjB,KAAK,EAAE,KAAK;;AAGd,mBAAoB;EAClB,KAAK,EAAE,KAAK;EACZ,MAAM,EAAE,IAAI;EACZ,OAAO,EAAE,IAAI;EACb,SAAS,EAAE,IAAI",
    "sources": Array [
      "./client/sass/core.scss",
      "./client/sass/main.scss",
    ],
    "version": "3",
  },
  "mapLocation": "single-line-inline.css",
}
`;

exports[`get map from file 4`] = `
Object {
  "map": Object {
    "file": "map-file-comment.css",
    "mappings": "AAAA,wBAAyB;EACvB,UAAU,EAAE,IAAI;EAChB,MAAM,EAAE,KAAK;EACb,OAAO,EAAE,IAAI;EACb,aAAa,EAAE,iBAAiB;EAChC,KAAK,EAAE,OAAkB;;AAG3B,wBAAyB;EACvB,OAAO,EAAE,IAAI;;ACTf,gBAAiB;EACf,UAAU,EAAE,IAAI;EAChB,KAAK,EAAE,MAAM;;AAGf,kBAAmB;EACjB,MAAM,EAAE,IAAI;EACZ,OAAO,EAAE,IAAI;EACb,UAAU,EAAE,KAAK;EACjB,aAAa,EAAE,GAAG;EAClB,KAAK,EAAE,KAAK;;AAEd,kBAAmB;EACjB,KAAK,EAAE,KAAK;;AAGd,mBAAoB;EAClB,KAAK,EAAE,KAAK;EACZ,MAAM,EAAE,IAAI;EACZ,OAAO,EAAE,IAAI;EACb,SAAS,EAAE,IAAI",
    "sources": Array [
      "./client/sass/core.scss",
      "./client/sass/main.scss",
    ],
    "version": "3",
  },
  "mapLocation": "map-file-comment.css.map",
}
`;

exports[`get map from file content 1`] = `
Object {
  "map": Object {
    "file": "map-file-comment.css",
    "mappings": "AAAA,wBAAyB;EACvB,UAAU,EAAE,IAAI;EAChB,MAAM,EAAE,KAAK;EACb,OAAO,EAAE,IAAI;EACb,aAAa,EAAE,iBAAiB;EAChC,KAAK,EAAE,OAAkB;;AAG3B,wBAAyB;EACvB,OAAO,EAAE,IAAI;;ACTf,gBAAiB;EACf,UAAU,EAAE,IAAI;EAChB,KAAK,EAAE,MAAM;;AAGf,kBAAmB;EACjB,MAAM,EAAE,IAAI;EACZ,OAAO,EAAE,IAAI;EACb,UAAU,EAAE,KAAK;EACjB,aAAa,EAAE,GAAG;EAClB,KAAK,EAAE,KAAK;;AAEd,kBAAmB;EACjB,KAAK,EAAE,KAAK;;AAGd,mBAAoB;EAClB,KAAK,EAAE,KAAK;EACZ,MAAM,EAAE,IAAI;EACZ,OAAO,EAAE,IAAI;EACb,SAAS,EAAE,IAAI",
    "sources": Array [
      "./client/sass/core.scss",
      "./client/sass/main.scss",
    ],
    "version": "3",
  },
  "mapLocation": "multi-line-inline.css",
}
`;

exports[`get map from file content 2`] = `
Object {
  "map": null,
  "mapLocation": "map-file-comment.css.map",
}
`;

exports[`get map from file content 3`] = `
Object {
  "map": Object {
    "file": "map-file-comment.css",
    "mappings": "AAAA,wBAAyB;EACvB,UAAU,EAAE,IAAI;EAChB,MAAM,EAAE,KAAK;EACb,OAAO,EAAE,IAAI;EACb,aAAa,EAAE,iBAAiB;EAChC,KAAK,EAAE,OAAkB;;AAG3B,wBAAyB;EACvB,OAAO,EAAE,IAAI;;ACTf,gBAAiB;EACf,UAAU,EAAE,IAAI;EAChB,KAAK,EAAE,MAAM;;AAGf,kBAAmB;EACjB,MAAM,EAAE,IAAI;EACZ,OAAO,EAAE,IAAI;EACb,UAAU,EAAE,KAAK;EACjB,aAAa,EAAE,GAAG;EAClB,KAAK,EAAE,KAAK;;AAEd,kBAAmB;EACjB,KAAK,EAAE,KAAK;;AAGd,mBAAoB;EAClB,KAAK,EAAE,KAAK;EACZ,MAAM,EAAE,IAAI;EACZ,OAAO,EAAE,IAAI;EACb,SAAS,EAAE,IAAI",
    "sources": Array [
      "./client/sass/core.scss",
      "./client/sass/main.scss",
    ],
    "version": "3",
  },
  "mapLocation": "single-line-inline.css",
}
`;

exports[`get map from file content 4`] = `
Object {
  "map": null,
  "mapLocation": "map-file-comment.css.map",
}
`;

exports[`get original location from map 1`] = `
Object {
  "column": 0,
  "line": 9,
  "name": null,
  "originalCode": null,
  "originalFile": "client/sass/core.scss",
  "source": "client/sass/core.scss",
}
`;

exports[`joinSourceMaps 1`] = `
Object {
  "file": "map-file-comment.css",
  "mappings": "AAAA,wBAAyB;EACvB,UAAU,EAAE,IAAI;EAChB,MAAM,EAAE,KAAK;EACb,OAAO,EAAE,IAAI;EACb,aAAa,EAAE,iBAAiB;EAChC,KAAK,EAAE,OAAkB;;AAG3B,wBAAyB;EACvB,OAAO,EAAE,IAAI;;ACTf,gBAAiB;EACf,UAAU,EAAE,IAAI;EAChB,KAAK,EAAE,MAAM;;AAGf,kBAAmB;EACjB,MAAM,EAAE,IAAI;EACZ,OAAO,EAAE,IAAI;EACb,UAAU,EAAE,KAAK;EACjB,aAAa,EAAE,GAAG;EAClB,KAAK,EAAE,KAAK;;AAEd,kBAAmB;EACjB,KAAK,EAAE,KAAK;;AAGd,mBAAoB;EAClB,KAAK,EAAE,KAAK;EACZ,MAAM,EAAE,IAAI;EACZ,OAAO,EAAE,IAAI;EACb,SAAS,EAAE,IAAI",
  "names": Array [],
  "sources": Array [
    "client/sass/core.scss",
    "client/sass/main.scss",
  ],
  "version": 3,
}
`;
